### ISSUE TITLE
Converting standalone module("xxxxx", package.seeall) .lua files, to work with openLuup

### ISSUE THREAD

**Issue by nodecentral** [akbooer/openLuup#31]
Hi @akbooer 

As always thank you so much for your support, i’m a big fan of openLuup, and making more of a leap over to it from Vera, by porting all my standalone modules, which is a big migration piece, hence I was confused why it was not working…

To try and wrap all this up, once and for all.. 🤪

My existing standalone Vera modules (.lua) all begin with e.g. `module("xxpushover", package.seeall)` and it looks like for them to work in openLuup, I need to ..

(a) remove the opening `module ()` reference, and replace it with e.g. `local xxpushover = {}` 

_My module file consists of both local values, and global functions_ 

(b) the file must end it with a return statement e.g. `return xxpushover`

To make use of the global functions in the file,  I need to add a `require` statement in my startup in the format of `local xxpushover = require "xxpushover"`

I have done all of the above, reloaded luup, yet when i call the following

`xxpushover.push_link("mysecretapicode", "openLuup Home Link Test", "Testing 1 2 3", 1, 'http://192.168.102.107/', "openLuup Home") `

I still get the  same `attempt to index global 'xxpushover' (a nil value)` error ?

**Comment by nodecentral** [comment]
For additional clarity, I have a luup.log statement in each module to show it’s been read, so I know openLuup is seeing/reading the file..

I’ve also tried the following, (i) change all global functions in the file to local, and (ii) adding the longer return statement proposed before e.g. 

````
return {
  push_file = push_file,
  push_basic = push_basic,
  push_link = push_link,
  push_html = push_html,
  push_mono = push_mono,
  push_specific = push_specific,
}
````

And still get the same `attempt to index global 'xxpushover' (a nil value)`

**Comment by akbooer** [comment]
Sorry you're still having problems.

Can you confirm *where and how* you are trying to access these functions (scene Lua code, for example?)

email me one of your modified modules so that I can give it a go.

**Comment by nodecentral** [comment]
Hi @akbooer, my update to an earlier post overlapped with your response.

I’m testing it via the Luup Test option in openLuup (console), and just running this..

`xxpushover.push_link("mysecretapicode", "openLuup Home Link Test", "Testing 1 2 3", 1, 'http://192.168.102.107/', "openLuup Home")`

**Comment by nodecentral** [comment]
It’s not a complicated module , just a mix of local variables and global functions, I posted a simplified version in the other thread - > https://github.com/akbooer/MetOffice_DataHub/issues/1#issuecomment-2452664513

I’ve looked in the lua files you created in the openLuup folder, to try learn/replicate what I could, while mine pale in comparison, (I like how yours start with an ABOUT section etc. ) - so I’m scratching my head on this..

**Comment by akbooer** [comment]
Testing in AltUI or openLuup Lua Test?

**Comment by akbooer** [comment]
Indeed, I’ve just tried this.  It is a ‘feature’ of AltUI, that it does not run its test code in the same environment as the startup/scene code.  It can’t because it’s run within its own plug-in environment which has no access to that.

openLuup, on the other hand, does this correctly, and you should find that things work from its own Lua Test environment.

So, since your initial modifications, it’s probably actually worked all the time, and should be fine to use in any scene code.

A learning exercise for both of us, so that’s good anyway.

**Comment by akbooer** [comment]
FYI... I tested with this in the Startup code:
```
TEST = require "moduleTest"
```
This in moduleTest.lua:
```
local _M = {}

function _M.hello(...)
  return "hello", ...
end

return _M
```
and ran this in openLuup's Lua Test:
```
print(TEST.hello "there")
```
with this result:
```
––– 1 line ––– 0.0 ms –––

hello 	there
```

**Comment by nodecentral** [comment]
> Testing in AltUI or openLuup Lua Test?

Hi, all my testing has been in openLuup (via Console)

I avoid using ALTUI as I'm not a fan of it.

**Comment by nodecentral** [comment]
Hi @akbooer, all of my modules have action functions that require multiple values to be provided, 

For example, this xxpushover (notification) has a global function for each type of pushover notification template (mono, link etc.). And when I call each of them there is a slightly different set of requirements. Another would be my xxionos module which has functions to send smtp email.

If I take a step back, an entry in the StartUp like `xxpushover = require "xxpushover"` reads the xxpushover.lua file.

The xxpushover.lua file that's read by openLuup begins with `local xxpushover = xxpushover {}` which is setting up the creation of an xxpushover table which as openLuup reads it sees local and global items listed. An example of local ones would be `http`, `https` etc. A global item would be a function such as `push_mono(priority, target, message)` that calls local or other global values to work.

For me, calling `xxpushover.push_mono(priority, target, message)` via Lua Test should work. I've somehow got it to work twice so far, but then when's i try it again later it gives me an error ...

**Comment by akbooer** [comment]
If it’s working at least once, then things can’t be too bad.

The only error that may have been introduced is if the module, internally, makes calls to some of its own exported functions, whose names you have just changed by adding the module prefix.

Using the alternative method of construction the module table at the end would avoid this potential issue.

Otherwise, I have no insights to offer on this limited information.   Any errors in the log?

**Comment by nodecentral** [comment]
It’s weird … :-)

He’s another of the modules I’m trying to convert  xxionos.lua, trying to follow your guide.

````
local _M = {}

-- Michal Kottman, 2011, public domain
local socket = require 'socket'
local smtp = require 'socket.smtp'
local ssl = require 'ssl'
local https = require 'ssl.https'
local ltn12 = require 'ltn12'

function _M.sslCreate()
    local sock = socket.tcp()
    luup.log("sslcreate called")
	print("sslcreate called")
    return setmetatable({
        connect = function(_, host, port)
            local r, e = sock:connect(host, port)
            if not r then return r, e end
            sock = ssl.wrap(sock, {mode='client', protocol='tlsv1_2'})
            return sock:dohandshake()
        end
    }, {
        __index = function(t,n)
            return function(_, ...)
                return sock[n](sock, ...)
            end
        end
    })
end

function _M.sendMessage(subject, body)
    local msg = {
        headers = {
            to = 'Your Target <my_name@hotmail.co.uk>',
            subject = subject
        },
        body = body
    }

    -- finally send it
    local ok, err = smtp.send {
        from = '< my_name@nodecentral.co.uk>',
        rcpt = '< my_name@hotmail.co.uk>',
        source = smtp.message(msg),
        user = 'my_name@nodecentral.co.uk',
        password = 'passwords2123',
        server = 'smtp.ionos.co.uk',
        port = 465,
        create = sslCreate
    }
    if not ok then
        luup.log("Mail send failed", err) -- better error handling required
        print("Mail send failed", err) -- better error handling required
    end
end

-- xxionos.sendMessage("VeraPlus (Test Email) Restart Notification", "Hi, \n\nThis is the email body.\n\nMore information will be added later.\n\nKind Regards\nVera Home Control")

luup.log(" <<<< - xxionos module loaded - >>>>> ")

return _M

-- return {
--   sendMessage = sendMessage,
-- }
````

I have the following in StartUp 

`xxionos = require "xxionos"`

And when I run the following it in LuaTest3 

`xxionos.sendMessage("VeraPlus (Test Email) Restart Notification", "Hi, \n\nThis is the email body.\n\nMore information will be added later.\n\nKind Regards\nVera Home Control")` 

it gives me this error.

````
––– 1 line ––– 0.1 ms –––
ERROR: [string "LuaTestCode3"]:1: attempt to index global 'xxionos' (a nil value)
````
Here is my start up log
````
2024-11-03 10:59:42.615   openLuup.init:: running _openLuup_STARTUP_
2024-11-03 10:59:42.615   luup_log:0:  <<<< - Started Loading Start Up Code - >>>>> 
2024-11-03 10:59:42.616   luup.register_handler:: global_function_name=rbLuaTest, request=LuaTest
2024-11-03 10:59:42.616   luup.register_handler:: global_function_name=lasttrip, request=LastTrip
2024-11-03 10:59:42.617   luup_log:0:  <<<< - xxpushover module loaded - >>>>> 
2024-11-03 10:59:42.617   luup_log:0:  <<<< - xxionos module loaded - >>>>> 
2024-11-03 10:59:42.617   luup_log:0:  <<<< - Finished Loading Start Up Code - >>>>> 
2024-11-03 10:59:42.617   openLuup.init:: init phase completed
````

UPDATE : I just did another reload and tried it again, this time LuaTest3 only showed this, no other information 
````
––– 0 lines ––– 0.7 ms –––
````
Hunting the logs, there is no record of the command I sent, but I did find this.
````
2024-11-03 11:01:38.312   openLuup.server:: POST /data_request?id=XMLHttpRequest&action=submit_lua&codename=LuaTestCode3 HTTP/1.1 tcp{client}: 0x55e5f9b97368
2024-11-03 11:01:48.385   closed:0: Mail send failed
````
So what is seen in the logs is not shown in the testing output window, which make it a little more difficult to troubleshoot, especially as I have been searching the log for anything with an `xx` which is my prefix for modules.

**Comment by akbooer** [comment]
This is not anything to do with openLuup itself.

I’d suggest that this may be a socket library / openSSL version issue.

**Comment by nodecentral** [comment]
Sorry, that was not the intent of my last post, it was just another example and a suggestion to consider including the command that’s sent too when testing via Lua Test, to help make the log entry more complete. (As the console testing result window does not show it)

**Comment by nodecentral** [comment]
The xxpushover module has progressed a little, StartUp has this

`xxpushover = require "xxpushover"`

My xxpushover.lua file (module) starts with 

`local xxpushover = {}`  and ends with `return xxpushover`

And now when sending the following request

````
xxpushover.push_mono("oeinfiowenfiwnfiuwnfiuwnfiuwenf", "Vera Home Mono Test", "Testing Mono 1 2 3", 0) 
````
It returns the following error
````
ERROR: [string "LuaTestCode2"]:1: attempt to call field 'push_mono' (a nil value)
````
Now, I’ve tried the `push_mono` function, as both a local and a global function but the error is the same? The fact I’m not getting an error on the module itself, it suggests that the xxpushover module is being found/seen, which is good 🎉 - so, this is perhaps where I’m getting lost …

Should ALL functions in the module be local, global and should they be module name prefixed. E.g..

````
function push_mono(apptoken, title, message, priority)
````
````
local function push_mono(apptoken, title, message, priority) 
````
Or
````
function xxpushover.push_mono(apptoken, title, message, priority)
````
````
local function xxpushover.push_mono(apptoken, title, message, priority) 
````

**Comment by akbooer** [comment]
If you want a function to be visible outside the module then, using this idiom, it should be prefixed by the module name.

You really need to take a step bck and understand the basics of Lua modules a bit more clearly.  But I know that reading the docs is not something that you ever enjoyed doing much ;-)

**Comment by nodecentral** [comment]
You know me so well :-) 

Sadly being dyslexic with an attention deficit disorder, documents are not easy for me, god knows I try - yet, I can find myself getting quite focussed trying to work something out, and try to ride that wave as much as I can when I do..

You can perhaps relate this to someone who learns the piano, but can’t read music :-) - I never understood why I struggled, nor perhaps did I like to admit it, but as I get older - hell I can’t be embarrassed by it anymore …

What’s weird with this, is I can get a module to work in openLuup a few times, , but then after a while it can suddenly not even see the module.. Now,  my ability to work between the different logs is not helping me, as my Lua test appears in the main log, yet any module load errors are in the startup log, and I don’t always think to look in there everytime, especially as I assume my changes are not breaking changes. 

So to sum up, it’s not you it’s all me. .. :-)

**Comment by akbooer** [comment]
No need for any embarrassment.

Does something, somewhere, set the global xxpushover to nil?

You can always check the content of a module by using, for example:

```
print(pretty(xxpushover))
```
